zookeeper/README.md

```diff
@@ -0,0 +1,94 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+implied. See the License for the specific language governing
+permissions and limitations under the License. See accompanying
+LICENSE file.
+-->
+
+## Quick Start
+
+This section describes how to run YCSB on ZooKeeper.
+
+### 1. Start ZooKeeper Server(s)
+
+### 2. Install Java and Maven
+
+### 3. Set Up YCSB
+
+Git clone YCSB and compile:
+
+    git clone http://github.com/brianfrankcooper/YCSB.git
+    # more details in the landing page for instructions on downloading YCSB(https://github.com/brianfrankcooper/YCSB#getting-started).
+    cd YCSB
+    mvn -pl site.ycsb:zookeeper-binding -am clean package
+
+### 4. Provide ZooKeeper Connection Parameters
+
+Set connectString, sessionTimeout, watchFlag in the workload you plan to run.
+
+- `zookeeper.connectString`
+- `zookeeper.sessionTimeout`
+- `zookeeper.watchFlag`
+  * A parameter for enabling ZooKeeper's watch, optional values:true or false.the default value is false.
+  * This parameter cannot test the watch performance, but for testing what effect will take on the read/write requests when enabling the watch.
+
+      ```bash
+      ./bin/ycsb run zookeeper -s -P workloads/workloadb -p zookeeper.connectString=127.0.0.1:2181/benchmark -p zookeeper.watchFlag=true
+      ```
+
+Or, you can set configs with the shell command, EG:
+
+    # create a /benchmark namespace for sake of cleaning up the workspace after test.
+    # e.g the CLI:create /benchmark
+    ./bin/ycsb run zookeeper -s -P workloads/workloadb -p zookeeper.connectString=127.0.0.1:2181/benchmark -p zookeeper.sessionTimeout=30000
+
+### 5. Load data and run tests
+
+Load the data:
+
+    # -p recordcount,the count of records/paths you want to insert
+    ./bin/ycsb load zookeeper -s -P workloads/workloadb -p zookeeper.connectString=127.0.0.1:2181/benchmark -p recordcount=10000 > outputLoad.txt
+
+Run the workload test:
+
+    # YCSB workloadb is the most suitable workload for read-heavy workload for the ZooKeeper in the real world.
+
+    # -p fieldlength, test the length of value/data-content took effect on performance
+    ./bin/ycsb run zookeeper -s -P workloads/workloadb -p zookeeper.connectString=127.0.0.1:2181/benchmark -p fieldlength=1000
+
+    # A Notice for the property:-p fieldcount
+    zookeeper does support the simple k-v data struct which the value is plaintext without any complex data struct,
+    so only pick up one field is enough for benchmark test, to avoid the cost of reading and deserializing the current record
+    in order to do an update of a subset of fields.So fieldcount is strictly constrained to 1,other value(like the below example) is invalid.
+    ./bin/ycsb run zookeeper -s -P workloads/workloadb -p zookeeper.connectString=127.0.0.1:2181/benchmark -p fieldcount=20
```

Comment:
- zookeeper does support the simple k-v data format and its value is simple without any complex data structure. The `setData ` api doesn't suport to update a subset of the znode data content. I cannot to read the znode with multi-fields, generate the new view of  multi-fields, then update the record in the method `update`  that will cause the measure of write opertion inaccurate(mixed with a read opertion).
- I constrained `fieldcount `to 1 because I want to insert/read/update a record with one field(In fact, for the zk benchmark ,we don't care about the `fieldcount `, we care about how the data length takes effect on the performance). If allowing for a configurable number of fields, I'll insert/read with 10 fields and update that znode to 1 field. The data volume of read and write is unequal and a record will shrink. Is it OK in the context of YCSB framework?
- I'm glad to listen to your insight :)

Reply:
> I cannot to read the znode with multi-fields, generate the new view of multi-fields, then update the record in the method update that will cause the measure of write opertion inaccurate(mixed with a read opertion).

Having to do the read in order to do the correct write is exactly what the benchmark in this case should reflect for an update. The update benchmark is specifically meant to show the difference between when an underlying datastore can do partial updates to a record and when it needs to read the complete record to update a field.

In order to avoid the read, either ZK would have to have some notion of a complex data type or you would have to structure record storage for this application so that a given znode was only the content of 1 field.

Given the current binding implementation here, where a znode corresponds to a single record, you will need to read the underlying znode before updating the set of fields passed for the update call.